mangoPURE/data/transforms.py

```diff
@@ -0,0 +1,280 @@
+        """
+        self.noises_dataset = noises_dataset
+        self.snrs = snrs
+        self.no_noise_coef = no_noise_coef
+
+    def __call__(
+            self,
+            example: MixedExample,
+    ) -> MixedExample:
+        """
+        Adds to the audio random noise from dataset at a random place
+        """
+        random_pos = random.randint(0, len(self.noises_dataset) - 1)
+        noise_example = self.noises_dataset[random_pos]
+        noise_audio = noise_example["audio"]["array"]
+        noise_len = noise_audio.shape[0]
+        audio_len = example.audio.shape[0]
+
+        if noise_len > audio_len:
+            noise_audio = noise_audio[:audio_len]
+            noise_len = audio_len
+
+        noise_coef = calc_audio_adjustment_coef(example.audio, noise_audio, self.snrs, self.no_noise_coef)
+
+        pos_start = random.randint(0, audio_len - 1)
+        if pos_start + noise_len <= audio_len:
+            example.audio[pos_start:pos_start + noise_len] += noise_coef * noise_audio
+            start, length = pos_start, noise_len
+        else:
+            end_len = audio_len - pos_start
+            start_len = noise_len - end_len
+            if start_len > end_len:
+                example.audio[:start_len] += noise_coef * noise_audio[-start_len:]
+                start, length = 0, start_len
+            else:
+                example.audio[-end_len:] += noise_coef * noise_audio[:end_len]
+                start, length = pos_start, end_len
+
+        if example.noises_info is None:
+            example.noises_info = []
+        example.noises_info.append(SegmentInfo(
+            start=start,
+            length=length,
+            class_id=noise_example["noise_id"]
+        ))
+
+        return example
+
+
+class AddSeveralRandomNoiseSegments(AddRandomNoiseSegment):
+    def __init__(
+            self,
+            noises_dataset: HfDataset,
+            snrs: tuple[float] = (10.0, 15.0, 20.0),
+            no_noise_coef: bool = True,
+            min_noises: int = 2,
+            max_noises: int = 4,
+    ):
+        """
+        :param min_noises: min noises segments in audio
+        :param max_noises: max noises segments in audio
+        see also AddRandomNoiseSegment
+        """
+        super().__init__(noises_dataset, snrs, False)
+        self.no_noise_coef = no_noise_coef
+        self.min_noises = min_noises
+        self.max_noises = max_noises
+
+    def __call__(
+            self,
+            example: MixedExample
+    ) -> MixedExample:
+        """
+        The same as AddRandomNoiseSegment, but it generates
+        several (random number) noises in the audio
+        """
+        example_helper = MixedExample(
+            audio=torch.zeros(example.audio.shape[0])
+        )
+        noises_num = random.randint(self.min_noises, self.max_noises)
+        for i in range(noises_num):
+            example_helper = super().__call__(example_helper)
```

Comment:
won't it add every new noise with broken snr? because we add noises one after one and recalculate snr each time - for last noise this nor will be ratio of this noise to all previous noises, and for the first one it will be 1. And then we calculate snr again...

Reply:
Nope, because (look line 254) I call the constructor of the super class with no_noise_coef=False.
Therefore this code just adds all noises and merges them in the end with the main audio

Reply:
😱